frame/staking/src/pallet/impls.rs

```diff
@@ -876,22 +876,19 @@ impl<T: Config> ElectionDataProvider for Pallet<T> {
 	}
 
 	fn voters(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<VoterOf<Self>>> {
-		// This can never fail -- if `maybe_max_len` is `Some(_)` we handle it.
 		let voters = Self::get_npos_voters(maybe_max_len);
-		debug_assert!(maybe_max_len.map_or(true, |max| voters.len() <= max));
-
+		if maybe_max_len.map_or(false, |m| voters.len() > m) {
+			defensive!("get_npos_voters is corrupt");
+		}
 		Ok(voters)
 	}
 
 	fn targets(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<T::AccountId>> {
-		let target_count = Validators::<T>::count();
-
-		// We can't handle this case yet -- return an error.
-		if maybe_max_len.map_or(false, |max_len| target_count > max_len as u32) {
-			return Err("Target snapshot too big")
+		let targets = Self::get_npos_targets(maybe_max_len);
+		if maybe_max_len.map_or(false, |m| targets.len() > m) {
```

Comment:
If you take my suggestion: https://github.com/paritytech/substrate/pull/10944/files#r816353287, then we should also make sure to check that fallback the upper bound respected if `maybe_max_len` is `None`